### ISSUE TITLE
Add an object-oriented API

### ISSUE THREAD

**Issue by dsuch** [dsuch/pymqi#173]
This is just an idea - perhaps to be implemented one day, possibly by someone in need of it and perhaps as a separate module or even its own project on top of PyMQI.

As it stands, PyMQI rather closely follows the C-level MQ calls, which is likely very good, but, from that, it follows that it is not particularly object-oriented.

For instance, in ticket #171, @salapat11 asked about the way to read in PutTime and PutDate from a message got from a queue.

A natural approach would be something like:

```python
message = queue.get()
put_time = message.md.PutTime
put_date = message.md.PutDate

data = message.data

# Process data along with its timestamp here
...
```

This would still require one to know that there is a thing in IBM MQ such as MQMD whose attributes are useful in various situations but at least there would be a single 'message' object to get access to everything through.

Instead of it, today, one basically sends MQMD or other MQ structures through references (directly mapping to pointers in C). This works but it is not very obvious unless one primarily works with C which is less likely now than at the time when PyMQI's API was first devised.

Perhaps it would be good to add a new object-oriented API a part of which would include the above - I leave it as an idea here.

**Comment by Skyler-Altol** [comment]
From a purely python point of view, it certainly would be more "pythonic" to have an OOP API, rather than essentially just wrapping raw C. While this'll likely cause a performance hit (though depending on implementation, may be large or small), I believe the benefits would outweigh the issues. OOP also allows static discovery of what methods can be applied to a given object, which is immensely useful for developers of all types, (amateurs, hobbyists, and professionals). I can't tell you how many times I've used this feature to figure out what I can do to an object without having to immediately consult documentation.

**Comment by dsuch** [comment]
Yes, code-completion is a great thing and it would be very nice to be able to at least code-complete all the various MQ structures, like ```msg.mqm<tab>.PutD<tab>``` if one already knows what to look for or, like you said, to discover what is possible, what is available. That could be the first step.

Generating Python classes from C structures does not sound hard but there are various places where for quite unknown reasons some padding needs to be applied for things to work. e.g. consult the pymqi.CD class and the last comment in its ```__init__``` method which adds 4 bytes of padding - this kind of scenarios will not be covered if classes are auto-generated without any customisation afterwards.

Another approach could be simply to rewrite the classes, I mean CD, GMO, PMO etc., and have them list all their attributes explicitly. This would easy and it is in fact what we do today anyway, except that today we do not have code-completion.

```python
class CD(:
  __slots__ = ('ChannelName', 'Version',...)

  def __init__(self):
    self.ChannelName = Bytes(20)
    self.Version = Long(CMQXC.MQCD_VERSION_6)
    ...
```

Where:

```python

classs Serialiser(object):
  # Serialisation goes here

class Bytes(bytes, Serializer):
  # Call serialisation methods here

class Long(int, Serialiser):
  # Call serialisation methods here
```

@Skyler-Altol - I am sure you understand what I mean. You said that you were a skilled Python programmer and you were now getting started with IBM MQ. Perhaps this is something that you would like to implement?

That would surely give you an opportunity to familiarise yourself with many corners of IBM MQ while being able to utilise your Python expertise. If you could start with just the base Serialiser class plus, say, CD and GMO, we could see if the idea is feasible on a small scale first.

**Comment by Skyler-Altol** [comment]
@dsuch unfortunately at this time I am working close to 11 hour days at my normal job, so I have no capacity to assist with building functionality in this API, at this time. My day job is a hybrid of MQ Administrator / Automation Engineer, hence I have extensive experience with administrating MQ itself, but little to no experience with actually building applications or programming against MQ. 

With that being said, understanding the application side would greatly assist with administration duties when issues occur. Should my time in the future become freed up, I may look into this.

**Comment by dsuch** [comment]
Sure, I understand it. Do stay around, though, if time permits. It would be great to have you here.

As for the parsing of MQSC responses - I did something like that long time ago, around MQ 7.x I think, using the PyParsing library, which you may already know - https://pypi.org/project/pyparsing/

At that time I just needed it in a limited scope for a given project but it should be easy to make it work with arbitrary commands and responses.

**Comment by Skyler-Altol** [comment]
@dsuch It has been a while good sir. Work is still as busy as ever, but I pretty much turned around and told them "nope, not doing these crazy hours anymore" when they're not paying for it. I've recently started toying around with the C API that this library is based on, and as of yesterday, have a basic C application that connects and opens queues. I like the idea of rewriting the classes to list their attributes explicitly, and will be syncing my fork and starting to toy around with that. There are a few other things that the C API has that I've noticed that PyMQI doesn't that I'm hoping to also stick in here, the biggest one being the application name feature that came in 9.1.2 (this makes it a lot easier for MQ administrators to identify what an application is).

https://www.ibm.com/docs/en/ibm-mq/9.1?topic=mqcno-applname-mqchar28
